### PR TITLE
Allow type as argument to throwerrors

### PR DESCRIPTION
**Pull request by MichaelHatherly**
~~Comments and docstrings probably need adjusting, I'll do that this afternoon if I have time.~~

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -133,7 +133,36 @@ end
...
+        return 42
+    end
+    @test c.error
+    @test c.output == "test\n"
+    @test c.value isa MethodError
+
     # Invalid throwerrors values
     @test_throws DomainError iocapture(()->nothing, throwerrors=:foo)
-    @test_throws TypeError iocapture(()->nothing, throwerrors=42)
+    @test_throws DomainError iocapture(()->nothing, throwerrors=42)
```
Comment (MichaelHatherly) [inline test/runtests.jl]: Are we bothered about this change at all?

File: src/IOCapture.jl [hunk 2]
```diff
@@ -49,11 +50,16 @@ It is also possible to set `throwerrors = :interrupt`, which will make `iocaptur
...
-    end
+function iocapture(f; throwerrors=Any)
+    # `throwerrors` is set to one of `true`, `false`, `:interrupt`, or a
+    # subtype of `ErrorException`, or a `Union` of error subtypes. Here we
+    # convert the first three choices to types instead, as:
+    #
+    #   - `true` -> `Any`,
+    #   - `false` -> `Union{}`,
+    #   - `:interrupt` -> `InterruptException`.
+    throwerrors = rewrite_error_argument(throwerrors)
```
Comment (MichaelHatherly) [inline src/IOCapture.jl]: Ain't terribly type-stable, but considering what else usually gets called in `iocapture` I doubt it's of any concern.

**Comment by MichaelHatherly** [comment]
Worth deprecating `true`, `false`, and `:interrupt` or shall we just leave them as is? I'm in favour of just leaving them in since it doesn't really make things any more complicated.

**Comment by mortenpi** [comment]
I would just get rid of `true`/`false`/`:interrupt` since they are duplicating the `::Type`-based interface, unless you think it would be useful for the user if we keep them? It's fine that it would be a breaking change -- I intentionally tagged this as 0.1, so that it would be possible to iterate on the design a little if need be.

File: test/runtests.jl (hunk 1 above)
Comment (mortenpi) [reply test/runtests.jl]: I would keep the test, but I wouldn't consider the exception type to be part of the public API, if that's what you mean.

File: src/IOCapture.jl [hunk 3]
```diff
@@ -16,10 +16,19 @@ in the terminal. Returns an object with the following fields:
...
+  by evaluating `f`. Setting it to `true` has the same effect as `Any`.
+
+  Setting it to `Union{}` will capture all thrown exceptions and return them
+  via the `.value` field, and will also set `.error` and `.backtrace`
+  accordingly. Setting it to `false` also has this effect.
+
+  To throw on a subset of possible exceptions pass the exception type instead,
+  such as `InterruptException`. If multiple exception types may need to be
+  thrown then pass a `Union{...}` of the types. Passing `:interrupt` has the
+  same effect as using `InterruptException`.
```
Comment (mortenpi) [inline src/IOCapture.jl]: I would emphasize here that the `InterruptException` case is the one you usually want, e.g. by changing the order of the paragraphs.

**Comment by MichaelHatherly** [comment]
> I would just get rid of true/false/:interrupt since they are duplicating the ::Type-based interface, unless you think it would be useful for the user if we keep them? It's fine that it would be a breaking change -- I intentionally tagged this as 0.1, so that it would be possible to iterate on the design a little if need be.

Either way is fine with me, shall I just drop them in this PR?

**Comment by mortenpi** [comment]
> shall I just drop them in this PR?

If you don't mind then yeah :slightly_smiling_face:

**Comment by MichaelHatherly** [comment]
Also makes default `Exception` rather than `Any`.

**Comment by mortenpi** [comment]
> Also makes default `Exception` rather than `Any`.

I would argue against that, because while you probably shouldn't, you can `throw` objects that are not subtypes of `Exception`:

```
julia> struct Foo end

julia> try
           throw(Foo())
       catch e
           @show e isa Exception
       end
e isa Exception = false
false
```

**Comment by MichaelHatherly** [comment]
Not bothered either way :) Bit weird that it throws anything though, never come across that in any real code.

**Comment by mortenpi** [comment]
LGTM, thanks a lot!

**Pull request merged**